File: javascript/Kookbook.py
from __future__ import with_statement

import sys, os, re, glob

package   = 'oktest'
release   = prop('release', '0.2.0')
copyright = 'copyright(c) 2010-2011 kuwata-lab.com all rights reserved'
license   = 'MIT License'

kookbook.default = "test"


node_path = os.environ.get('NODE_PATH')
if not node_path:
    node_path = os.path.dirname(os.path.abspath(__file__))
    os.environ['NODE_PATH'] = os.path.join(node_path, 'lib')


nodejs = "node --strict_mode"


def dist_dir():
    return "dist/oktest-" + release


vs_home = os.getenv('VS_PATH', '').split(':')[0]
nodejs_executables = [
    ('0.4', vs_home + '/node/0.4.11/bin/node'),
    ('0.5', vs_home + '/node/0.5.10/bin/node'),
    ('0.6', vs_home + '/node/0.6.3/bin/node'),
]


class test(Category):

    @recipe
    def default(c):
        """do test"""
        #system(c%"$(nodejs) math_test.js")
        with chdir("test"):
            for fname in glob.glob("*_test.js"):
                system(c%"$(nodejs) $(fname)")

    @recipe
    def all(c):
        """do test on Node.js 0.4, 0.5. and 0.6"""
        with chdir("test"):
            for ver, bin in nodejs_executables:
                print("******************** Node.js %s ********************" % ver)
                for fname in glob.glob("*_test.js"):
                    system(c%"$(bin) $(fname)")


text_files = [
    'README.txt', 'CHANGES.txt', 'MIT-LICENSE', 'Kookbook.py', 'package.json',
]


@recipe
def dist(c):
    """create 'dist-x.x.x' directory and copy files to it"""
    dir = dist_dir()
    rm_rf(dir)
    mkdir_p(dir)
    #
    store(text_files, dir)
    store("lib/*.js", "doc/users-guide.html", "doc/docstyle.css", "test/*_test.js", dir)
    #
    #mkdir(c%"$(dir)/bin")
    #cp("lib/oktest.js", c%"$(dir)/bin/oktest.js")
    #system(c%"chmod 0755 $(dir)/bin/oktest.js")
    #
    replacer = [
        (r'\$Release:.*?\$',   r'$Release: %s $'   % release),
        (r'\$Copyright:.*?\$', r'$Copyright: %s $' % copyright),
        (r'\$License:.*?\$',   r'$License: %s $'   % license),
        (r'\$Release\$',   release),
        (r'\$Copyright\$', copyright),
        (r'\$License\$',   license),
    ]
    edit(c%'$(dir)/**/*', by=replacer)
    cp('Kookbook.py', dir)   # don't edit 'Kookbook.py'


@recipe
@ingreds('dist')
def publish(c):
    """upload package; you must do 'npm adduser' in advance"""
    with chdir(dist_dir()):
        system("npm publish .")


HOME = os.getenv('HOME')


if not os.path.isdir('doc'):
    mkdir('doc')


@recipe
@ingreds('doc/users-guide.html', 'doc/docstyle.css')
def doc(c):
    """create documents under 'doc' directory"""
    pass


@recipe('doc/*.html')
@ingreds('doc/$(1).txt')
def file_doc_html(c):
    base = os.path.basename(c.product).replace('.html', '')
    with chdir('doc'):
        system(c%'kwaser -t html-css -T $(base).txt > $(base).toc.html')
        system(c%'kwaser -t html-css    $(base).txt > $(base).html')
        rm_f(c%'$(base).toc.html')


@recipe('doc/docstyle.css')
@ingreds(HOME + '/src/kwaser/docstyle.css')
def file_doc_docstyle_css(c):
    rm_f(c.product)
    system(c%'ln $(ingred) $(product)')

File: perl/Kookbook.pl
use strict;
use warnings;

my $project   = prop('project', 'Oktest');
my $release   = prop('release', '0.0102');
my $copyright = 'copyright(c) 2010-2011 kuwata-lab.com all rights reserved';
my $license   = 'MIT License';

$kook_default = 'test';

my @perl_bin_paths = (
    '/opt/lang/perl/5.8.9/bin',
    '/opt/lang/perl/5.10.1/bin',
    '/opt/lang/perl/5.12.4/bin',
    '/opt/lang/perl/5.14.2/bin',
);

recipe 'test', {
    desc   => 'do test',
    spices => ["-a: test with Perl 5.8, 5.10, 5.12, and 5.14"],
    method => sub {
        my ($c, $opts) = @_;
        if ($opts->{a}) {
            for my $binpath (@perl_bin_paths) {
                my $ver = ($binpath =~ /5(?:\.\d+)+/ && $&);
                print "#---- Perl $ver --------------------\n";
                sys "$binpath/prove t";
            }
        }
        else {
            sys 'prove t';
            #sys 'bin/oktest.pl examples';
        }
    }
};

recipe 'examples', {
    desc   => 'run examples',
    method => sub {
        sys 'bin/oktest.pl -sv examples';
    }
};

recipe 'README', {
    ingreds => ['lib/Oktest.pm'],
    method  => sub {
        sys 'pod2text lib/Oktest.pm > README';
    }
};

recipe 'package', {
    desc    => 'create package',
    ingreds => ['dist'],
    method  => sub {
        my $dir = "dist/$project-$release";
        cd $dir, sub {
            sys 'perl Makefile.PL';
            sys 'make';
            #sys 'make disttest';
            sys 'make dist';
        };
        mv "$dir/$project-$release.tar.gz", '.';
    }
};

my @text_files = qw(Changes Kookbook.pl MANIFEST MIT-LICENSE Makefile.PL README);

recipe 'dist', {
    desc    => 'copy and edit files into dist directory',
    ingreds => ['README'],
    method  => sub {
        ## create directory
        my $dir = "dist/$project-$release";
        rm_rf $dir if -d $dir;
        mkdir_p "dist/$project-$release";
        ## copy files
        store @text_files, $dir;
        store 'bin/*', 'lib/**/*', 't/**/*', 'examples/**/*', $dir;
        ## create MANIFEST file
        cd $dir, sub {
            #sys 'make -f ../../Makefile manifest';
            #rm 'MANIFEST';
            #sys 'perl "-MExtUtils::Manifest=mkmanifest" -e mkmanifest 2>/dev/null';
            sys "find . -type f | sed -e 's=^\\./==g' > MANIFEST";
            cp 'MANIFEST', '../..';
            #rm 'MANIFEST.bak' if -f 'MANIFEST.bak';
        };
        ## edit files
        cd $dir, sub {
            edit '**/*', sub {
                s/\$Release\$/$release/eg;
                #s/\$Copyright\$/$copyright/eg;
                #s/\$License\$/$license/eg;
                s/\$Release: .*? \$/"\$Release: $release \$"/eg;
                #s/\$Copyright: .*? \$/"\$Copyright: $copyright \$"/eg;
                #s/\$License: .*? \$/"\$License: $license \$"/eg;
                $_ || '-';
            };
            chmod 0755, $_ for glob('bin/*');
        };
        ##
    }
};
